.github/ISSUE_TEMPLATE/test_data_removal.md

```diff
@@ -0,0 +1,27 @@
+---
+name: Test data removal request
+about: Request removal of your content from our test dataset
+title: ''
+assignees:
+ - jayaddison
+
+---
+
+**Test data removal**
+
+<!-- NOTE: Please **do not** include your email address in this bugreport, as the bugreport will be public and doing so could expose your email address to spammers. -->
+
+<!-- Please describe the test data that you would like to be removed below -->
+
+Please could you remove the following `recipe-scrapers` test data files:
+
+  - e.g. `test_example_scraper.testhtml`
+
+
+The following statements apply to my request:
+
+  - [ ] I am the owner of content contained within those files, and would like them to be removed.
+
+  - [ ] I would also like them to be erased from all versions of the source control history.
+
+<!-- NOTE: the second request may be difficult for us to fulfill; other copies of the source control exist outside of our control, and rewriting the code history makes is potentially disruptive for users and developers of this library.  however, if you require this, we may be able to accommodate your request under some circumstances -->
```

Comment:
Rewriting commit history to filter out content from all versions/branches necessarily causes `git` commit IDs (SHA1 hashes) to change.  This could cause some concern for developers/users who want to assure that they're running the authentic, original versions of our code.

To prepare for this possibility, we could create SHA256 hashes of the contents of `git archive` for all of the published release tags in this repository.  Since those are cryptographic hashes, they don't include the content of removed requests.

I'm not sure where we would publish that index yet -- and we might need to append to it, but it is an idea to reassure and continue to provide integrity to users/developers.

Reply:
The tricky aspect of this is that we only want to hash the parts of each `git` tree that were included in the Python package produced as output.  Because otherwise those hashes aren't independently verifiable without access to the original, pre-filtered repository history.

...and an additional detail to keep in mind when implementing this is that the path to the test data changed with v6.1.0 (related to 9ae46af6a53d766ab5927fa1957c2ff572e4b100).

Reply:
Maybe not so tricky if we simply filter the `git archive` to the path `recipe_scrapers`.  That's always been the base of the Python code in this library, and that's been consistent throughout all the release (so it should, I think, be straightforward to verify).

Therefore I initially propose to do this using a command something like:

```sh
$ for i in $(git tag -l | sort -n); do echo $i $(git archive --format=tar $i^{tree} recipe_scrapers | sha256sum); done;
```

I'll plan to run that on at least two separate computers to confirm whether the results are the same.  The first ten lines of output on my machine here are:

```
5.0.0 c9f6d8fb7c845711b1eb001f75c44e4ac5b3f6c2e998cbf61f874ba26436d239 -
5.0.1 c9f6d8fb7c845711b1eb001f75c44e4ac5b3f6c2e998cbf61f874ba26436d239 -
5.1.0 9b185f74d99cdd11ee5eba6ee5eb50727547d6caac7be0f68630c8bcbde20dff -
5.14.0 8cb7f192b20a86eee0d9c012d70b56f919c05352b6065e64ee8e0f3c67c2245c -
5.15.0 77e510d8d44434568def47c158047b17630bdd88669502833b211e100365da04 -
5.16.0 a4faa5510b5f306c64549064973c569af02c515d1b56c86c55e99f7e0878aa40 -
5.17.0 127b76d588794c92ec40002c1a467ea362674bd7f8c9328d9ec5c5305e867c1c -
5.18.0 c2ae7d4b99496be4a6c63a01894fab4d82744654ed8907bb46b05e6dbdabf41b -
5.2.0 4783ce2f16386f07893da6e1960fde1f9bdee425e4cf193eec7c0cb1f928bf50 -
5.2.1 57b1a8576af63b1c4ba9a473cb324b8b8b417519ae4a88f23ca00303d2b43072 -
```

(and as a partial sense-check, 5.0.0 to 5.0.1 did indeed not involve any edits to the `recipe_scrapers` dir inside this repository)

Reply:
> Therefore I initially propose to do this using a command something like:
> 
> ```shell
> $ for i in $(git tag -l | sort -n); do echo $i $(git archive --format=tar $i^{tree} recipe_scrapers | sha256sum); done;
> ```

That won't work for the `recipe_scrapers` directory pre-6.1.0, because the `tests` directory exists at `recipe_scrapers/tests` -- so any filtering of test data would result in different `git archive` output.

The best I've got so far is this updated, larger script:

```sh
for i in $(git tag -l | sort -n);
do
        git checkout $i > /dev/null 2>&1;
        # test files have always been excluded from the package build...
        # https://github.com/hhursev/recipe-scrapers/blob/93244196fd17b26f211a68d5ddc0ecdb67102370/MANIFEST.in#L5
        # https://github.com/hhursev/recipe-scrapers/blob/ef5024de3b39ca40755c797bf177583c56613886/MANIFEST.in#L6
        # https://github.com/hhursev/recipe-scrapers/blob/aacd321f96235e44e438202c76f7eceddf4aa193/pyproject.toml#L32
        # ...so we want to ensure they're omitted from these content hashes
        rm -rf recipe_scrapers/tests  # pre-6.1.0
        rm -rf tests  # post-6.1.0
        git add . && git commit -m 'ephemeral' --allow-empty > /dev/null 2>&1;
        echo $i $(tar --create --exclude=tests --mtime=0 --sort=name recipe_scrapers | sha256sum);
done;
```

This produces consistent output when I run it on the current copy of `recipe-scrapers.git` and also on a cloned copy of the repository that has all `.testhtml` files pruned from the entire history (`git-filter-repo --invert-paths --path-glob '*.testhtml' --use-base-name`).  Next, I'll attempt that on a separate computer.

Reply:
A few problems with that script:

  * It applies local UID and GID based on the system it runs on, making results difficult to verify independently.
  * It also adds some file permissions to the tarfile based on the local system (umask, I think), and those also make the results difficult to replicate.
  * The tag checkout and commit logic is needlessly complicated, and I think may introduce some timing problems.

An improved version:

```sh
for i in $(git tag -l | sort -n);
do
        git checkout --quiet .;
        git checkout --quiet $i;
        # test files have always been excluded from the package build...
        # https://github.com/hhursev/recipe-scrapers/blob/93244196fd17b26f211a68d5ddc0ecdb67102370/MANIFEST.in#L5
        # https://github.com/hhursev/recipe-scrapers/blob/ef5024de3b39ca40755c797bf177583c56613886/MANIFEST.in#L6
        # https://github.com/hhursev/recipe-scrapers/blob/aacd321f96235e44e438202c76f7eceddf4aa193/pyproject.toml#L32
        # ...so we want to ensure they're omitted from these content hashes
        rm -rf recipe_scrapers/tests  # pre-6.1.0
        rm -rf tests  # post-6.1.0
        echo $i $(tar --create --exclude=tests --group=0 --mode='a=r' --mtime=0 --owner=0 --sort=name recipe_scrapers | sha256sum);
done;
```

This produces the same output for the first few tags (5.0.0, ...) when I run it on two separate computers.

```
5.0.0 52d3002ff1db532af2281c7e1a2e8d76a401100e6fff71ca8be6df45b472e0e3 -
5.0.1 52d3002ff1db532af2281c7e1a2e8d76a401100e6fff71ca8be6df45b472e0e3 -
5.1.0 d1e41a449f661c684557a0838ae51b770d591b75c99dc4fc8f480eec73c9665c -
5.14.0 90ca1bbd5491a5e0194330472a7d2adc45ab6717a7ead823da9d2168946ee358 -
5.15.0 d86d6754a82254a7141d5a0eea05ef2919eca28a39662777a98f639066d14c62 -
```

Reply:
This now seems fairly reliable (provided that the working directory tree is clean to begin with).  I've also been able to replicate the exact same hashes when comparing two released git tags to their published versions on PyPi using this method.  Not too surprising given that the library contains no compiled code.. but still reassuring.

Reply:
**tl;dr** - I think we should try very, very hard not to filter/remove content from git history.  But to prepare for the eventuality that we might need to, here's a method to confirm that the packages were built without the test data and that their code hasn't changed in git after the filtering.